service/worker/indexer/esProcessor.go

```diff
@@ -0,0 +1,265 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package indexer
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/olivere/elastic"
+	"github.com/uber-common/bark"
+	"github.com/uber/cadence/.gen/go/indexer"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/collection"
+	"github.com/uber/cadence/common/logging"
+	"github.com/uber/cadence/common/messaging"
+	"github.com/uber/cadence/common/metrics"
+	"time"
+)
+
+type (
+	// ESProcessor is interface for elastic search bulk processor
+	ESProcessor interface {
+		// Stop processor and clean up
+		Stop()
+		// Add request to bulk, and record kafka message in map with provided key
+		Add(request elastic.BulkableRequest, key string, kafkaMsg messaging.Message)
+	}
+
+	// ElasticBulkProcessor is interface for elastic.BulkProcessor
+	// (elastic package doesn't provide such interface that tests can mock)
+	ElasticBulkProcessor interface {
+		Start(ctx context.Context) error
+		Stop() error
+		Close() error
+		Stats() elastic.BulkProcessorStats
+		Add(request elastic.BulkableRequest)
+		Flush() error
+	}
+
+	// esProcessorImpl implements ESProcessor, it's an agent of elastic.BulkProcessor
+	esProcessorImpl struct {
+		processor     ElasticBulkProcessor
+		mapToKafkaMsg collection.ConcurrentTxMap // used to map ES request to kafka message. key: esDocID+visType
+		config        *Config
+		logger        bark.Logger
+		metricsClient metrics.Client
+	}
+)
+
+var _ ESProcessor = (*esProcessorImpl)(nil)
+
+const (
+	// retry configs for es bulk processor
+	esProcessorInitialRetryInterval = 200 * time.Millisecond
+	esProcessorMaxRetryInterval     = 20 * time.Second
+)
+
+// NewESProcessorAndStart create new ESProcessor and start
+func NewESProcessorAndStart(config *Config, client *elastic.Client, processorName string,
+	logger bark.Logger, metricsClient metrics.Client) (ESProcessor, error) {
+	p := &esProcessorImpl{
+		config: config,
+		logger: logger.WithFields(bark.Fields{
+			logging.TagWorkflowComponent: logging.TagValueIndexerESProcessorComponent,
+		}),
+		metricsClient: metricsClient,
+	}
+
+	processor, err := client.BulkProcessor().
+		Name(processorName).
+		Workers(config.ESProcessorNumOfWorkers()).
+		BulkActions(config.ESProcessorBulkActions()).
+		BulkSize(config.ESProcessorBulkSize()).
+		FlushInterval(config.ESProcessorFlushInterval()).
+		Backoff(elastic.NewExponentialBackoff(esProcessorInitialRetryInterval, esProcessorMaxRetryInterval)).
+		After(p.bulkAfterAction).
+		Do(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	p.processor = processor
+	p.mapToKafkaMsg = collection.NewShardedConcurrentTxMap(1024, p.hashFn)
```

Comment:
I think it is fine to init with 1024 (each map is init with 32 since we use fixed 32 shards). Because writer to this map have `config.IndexerConcurrency` (default 1000) concurrency, while reader of this map have `config.ESProcessorNumOfWorker` (default 1) concurrency.